libs/ml/include/ml/ops/squeeze.hpp

```diff
@@ -0,0 +1,129 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "math/standard_functions/log.hpp"
+#include "ml/ops/ops.hpp"
+
+#include <cassert>
+
+namespace fetch {
+namespace ml {
+namespace ops {
+
+template <class T>
+class Squeeze : public fetch::ml::ops::Ops<T>
+{
+public:
+  using TensorType    = T;
+  using DataType      = typename TensorType::Type;
+  using SizeType      = typename TensorType::SizeType;
+  using VecTensorType = typename Ops<T>::VecTensorType;
+  using SPType        = OpSqueezeSaveableParams<T>;
+  using MyType        = Squeeze<TensorType>;
+
+  Squeeze() = default;
+
+  explicit Squeeze(SPType const &sp)
+    : Ops<T>(sp)
+  {}
+
+  ~Squeeze() override = default;
+
+  std::shared_ptr<OpsSaveableParams> GetOpSaveableParams() override
+  {
+    auto sp = std::make_shared<SPType>();
+    return sp;
+  }
+
+  std::shared_ptr<fetch::ml::ops::Ops<TensorType>> MakeSharedCopy(
+      std::shared_ptr<fetch::ml::ops::Ops<TensorType>> me) override
+  {
+    FETCH_UNUSED(me);
+    assert(me.get() == this);
+
+    auto copyshare = std::make_shared<MyType>(*this);  // calls default copy constructor of MyType
+
+    return copyshare;
+  }
+  /**
+   * elementwise Squeeze
+   * @param inputs vector containing one tensor which is the input tensor to Squeeze
+   * @return
+   */
+  void Forward(VecTensorType const &inputs, TensorType &output) override
+  {
+    assert(inputs.size() == 1);
+    assert(output.shape() == this->ComputeOutputShape(inputs));
+
+    output.Copy((*inputs.at(0)));
+    output.Squeeze();
+  }
+
+  /**
+   * elementwise Squeeze gradient is 1/x * error:
```

Comment:
if this is true, where is the division by x happening?

Reply:
Copy/paste error